### PR TITLE
Fix errors caused by newline and whitespace in folded string

### PR DESCRIPTION
**Pull request by setcy**
fixes: #41

## Test data:
input: `hello \n world`
output:
```yaml
>-
  hello 
   world
```

input: `hello \n\n world`
output:
```yaml
>-
  hello 
  
   world
```

input: `hello \n \n world`
output:
```yaml
>-
  hello 
   
   world
```

input: `hello \nstring\n world`
output:
```yaml
>-
  hello 

  string
   world
```

### CODE REVIEW

File: lib/src/strings.dart [hunk 1]
```diff
@@ -105,12 +105,31 @@ String _tryYamlEncodeFolded(String string, int indentation, String lineEnding) {
     result = '>-\n' + ' ' * indentation;
   }
 
+  /// If the line starts with a empty character (excluding indentation), the
+  /// newline character(\n) before and after the line will be remained.
+  var emptyBegin = false;
+  final replacedString =
+      trimmedString.replaceAllMapped(RegExp(r'\n(.*)'), (match) {
+    final nextLine = match.group(1)!;
```
Comment (jonasfj) [inline lib/src/strings.dart:113]: Will `.` not match `\n`, hence, `nextLine` is more than just the next line?

File: lib/src/strings.dart [hunk 2]
```diff
@@ -105,12 +105,31 @@ String _tryYamlEncodeFolded(String string, int indentation, String lineEnding) {
     result = '>-\n' + ' ' * indentation;
   }
 
+  /// If the line starts with a empty character (excluding indentation), the
+  /// newline character(\n) before and after the line will be remained.
```
Comment (jonasfj) [inline lib/src/strings.dart:109]: This comment needs some improvement

File: lib/src/strings.dart [hunk 3]
```diff
@@ -105,12 +105,31 @@ String _tryYamlEncodeFolded(String string, int indentation, String lineEnding) {
...
+  var emptyBegin = false;
+  final replacedString =
+      trimmedString.replaceAllMapped(RegExp(r'\n(.*)'), (match) {
```
Comment (jonasfj) [inline lib/src/strings.dart:112]: Would it be better to do `.split('\n')`, not sure. Just asking if we've considered it?

File: test/utils_test.dart [hunk 4]
```diff
@@ -412,6 +412,48 @@ a:
...
+
+        test('with whitespace newline string newline whitespace', () {
+          final doc = YamlEditor('hello');
+          final node = wrapAsYamlNode(
+            'hello \nstring\n world',
+            scalarStyle: ScalarStyle.FOLDED,
+          );
+          doc.update([], node);
+          expect(doc.toString(), equals('>-\n  hello \n\n  string\n   world'));
+        });
```
Comment (jonasfj) [inline test/utils_test.dart:455]: Maybe this is a separate issue, but might also be worth considering if we should address it here.

Are there strings that cannot be written as folded string?

Playing around I had a hard time writing ` T` or `\n T`

**Comment by setcy** [comment]
Hi @jonasfj, thanks for the review!

I have pushed a new commit that updates the comment and improves the code by using `.split('\n')`.

> Maybe this is a separate issue, but might also be worth considering if we should address it here.
> 
> Are there strings that cannot be written as folded string?
> 
> Playing around I had a hard time writing ` T` or `\n T`

For these conditions, we can use [block styles with indication indicators](https://yaml.org/spec/1.2.2/#8111-block-indentation-indicator) to implement it.
### condition 1:
input: ` hello`
output:
```yaml
>-3
   hello
```

### condition 2:
input: `\n hello`
output:
```yaml
>-3

   hello
```

### condition 3:
input: `      \n    \n  hello \n   `
output: 
```yaml
>-3
        
      
    hello 
     
```
note: Some yaml conversion tools could not correctly parse condition 3, they parsed the output as `\n\n hello`

**Pull request closed without merging**